### PR TITLE
feat(reviews): support Qodo issue-comment suggestions in review-handler

### LINKED ISSUE #112: feat(reviews): support Qodo issue-comment suggestions in review-handler
## Problem

Qodo's `/improve` workflow posts code suggestions as **PR issue comments** (not review threads). The review-handler only processes review threads via `pullRequestReviewThreads` GraphQL query, so these suggestions are completely invisible to the handler.

Example: https://github.com/myk-org/claude-code-config/pull/111#issuecomment-3876844528

### What Qodo issue comments look like

- Posted by `qodo-code-review[bot]` as issue comments
- Contain an HTML table with code suggestions
- Each suggestion has: category, file path + line range, diff (before/after), importance score (1-10)
- Suggestions include interactive checkboxes (`Apply / Chat`, `More`)

### Why the handler misses them

`myk-claude-tools reviews fetch` only queries `pullRequestReviewThreads` (GraphQL), which returns inline review comments. Qodo issue comments are a different GitHub API entity (`/issues/{number}/comments`).

### 1. Fetch Qodo issue comments

In `fetch.py`, after fetching review threads, also query PR issue comments and filter for `qodo-code-review[bot]` author. Parse the HTML table body to extract individual suggestions.

### 2. Parse suggestion structure

Each Qodo suggestion contains:
- **Category** (e.g., "best practice", "Enhancement")
- **File path + line range** (e.g., `myk_claude_tools/reviews/post.py [468-472]`)
- **Diff block** (before/after code)
- **Importance score** (1-10)
- **Suggestion text** (description of the change)

### 3. Map to existing thread structure

Convert each suggestion into the same JSON structure used for review threads:
- `source`: `"qodo"` (same category)
- `type`: `"issue_comment_suggestion"` (to distinguish from inline review)
- `comment_id`: the issue comment ID
- `suggestion_index`: index within the comment (since one comment has multiple suggestions)
- `path`, `line`, `body`, `priority` mapped from parsed data

### 4. Posting replies

Since these are issue comments (not threads), replying works differently:
- Cannot reply inline to individual suggestions
- Could reply to the issue comment with a summary of which suggestions were addressed/skipped
- Or post a new issue comment referencing the original

## Considerations

- Qodo also posts review threads for `/review` — those already work
- Only `/improve` suggestions come as issue comments
- The HTML table parsing needs to handle Qodo's specific format (may change between versions)
- Should we also support CodeRabbit issue comments? (CodeRabbit posts a summary comment too)

## Deliverables

- [ ] Research Qodo issue comment format stability (is the HTML structure consistent?)
- [ ] Add issue comment fetching to `fetch.py`
- [ ] Add HTML table parser for Qodo suggestion format
- [ ] Map parsed suggestions to review JSON structure
- [ ] Handle reply posting for issue comment suggestions
- [ ] Add tests for parsing and fetching
- [ ] Update `review-handler.md` documentation

### PR DESCRIPTION
**Pull request by myakove**
## Summary

- Add HTML parser (`qodo_parser.py`) for Qodo `/improve` and `/review` issue comment formats
- Extend `fetch.py` to auto-fetch Qodo issue comments and parse individual suggestions
- Extend `post.py` with batch posting of summary replies for issue comment suggestions
- Add idempotent `posted_at` tracking to prevent duplicate posts on retry

Closes #112

### New files
| File | Purpose |
|------|---------|
| `myk_claude_tools/reviews/qodo_parser.py` | Stdlib-only HTML parser for Qodo comment formats |
| `tests/test_qodo_parser.py` | 46 tests for parser |

### Modified files
| File | Changes |
|------|---------|
| `fetch.py` | `fetch_qodo_issue_comments()`, updated `run()` and `get_thread_key()` |
| `post.py` | `post_issue_comment()`, `build_issue_comment_reply()`, batch posting in `run()` |
| `test_get_all_reviews.py` | 6 new tests |
| `test_post_review_replies.py` | 10 new tests |
| `review-handler.md` | Phase 1 and Phase 6 documentation |
| `.gitignore` | Whitelisted new files |

## Design decisions

- Each parsed suggestion becomes a separate thread item (`type: "issue_comment_suggestion"`)
- Suggestions grouped by `issue_comment_id` for batch posting (one reply per Qodo comment)
- Composite dedup key `ic:{comment_id}:{suggestion_index}` in `get_thread_key()`
- No thread resolution (issue comments don't support GitHub's resolve concept)
- `posted_at` timestamps prevent duplicate posts on retry

## Test plan

- [x] All 544 tests pass (`uvx --with tox-uv tox`)
- [x] All 20 pre-commit hooks pass
- [ ] Manual test: `myk-claude-tools reviews fetch` on a PR with Qodo `/improve` comments
- [ ] Manual test: `myk-claude-tools reviews fetch` on a PR with Qodo `/review` comments

___

### **PR Type**
Enhancement

___

### **Description**
- Add Qodo issue comment suggestion parsing and batch posting

- Support `/improve` and `/review` comment formats with HTML parser

- Auto-fetch and categorize Qodo issue comments as separate suggestions

- Batch-post summary replies grouped by issue comment ID

___

### Diagram Walkthrough

```mermaid
flowchart LR
  QodoComments["Qodo Issue Comments<br/>/improve & /review"]
  Fetch["fetch_qodo_issue_comments()"]
  Parser["qodo_parser.py<br/>HTML parsing"]
  Suggestions["Individual Suggestions<br/>type: issue_comment_suggestion"]
  Batch["Batch Post<br/>per comment ID"]
  Reply["Issue Comment Reply<br/>Summary table"]
  
  QodoComments --> Fetch
  Fetch --> Parser
  Parser --> Suggestions
  Suggestions --> Batch
  Batch --> Reply
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>qodo_parser.py</strong><dd><code>Qodo comment HTML parser for suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

myk_claude_tools/reviews/qodo_parser.py

<ul><li>New stdlib-only HTML parser for Qodo <code>/improve</code> and <code>/review</code> comment <br>formats<br> <li> <code>parse_improve_comment()</code> extracts suggestions with category, diff, <br>importance, impact<br> <li> <code>parse_review_comment()</code> extracts focus areas from review guide section<br> <li> <code>_extract_path_from_link_text()</code> helper parses file paths and line <br>ranges</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-5a0f7db20783b0cdf11f870a3ef77712dd2b1e9d595436999b81218d8c2cb092">+332/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fetch.py</strong><dd><code>Fetch and parse Qodo issue comment suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

myk_claude_tools/reviews/fetch.py

<ul><li>Add <code>fetch_qodo_issue_comments()</code> to fetch and parse Qodo issue comments<br> <li> Each parsed suggestion becomes separate thread item with <code>type: </code><br><code>"issue_comment_suggestion"</code><br> <li> Update <code>get_thread_key()</code> to use composite key <br><code>ic:{comment_id}:{suggestion_index}</code><br> <li> Integrate Qodo issue comment fetching into <code>run()</code> workflow<br> <li> Update user message for issue comment URL handling</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-46f5ef6a1dc4b99c18488f637a1344593c06196e5341c75e606007d41240df5a">+71/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>post.py</strong><dd><code>Batch post replies for issue comment suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

myk_claude_tools/reviews/post.py

<ul><li>Add <code>post_issue_comment()</code> to post new issue comments on PRs<br> <li> Add <code>build_issue_comment_reply()</code> to format suggestions as markdown <br>table<br> <li> Implement batch posting logic for issue comment suggestions grouped by <br>comment ID<br> <li> Track <code>posted_at</code> timestamps to prevent duplicate posts on retry<br> <li> Skip thread resolution for issue comment suggestions (no thread_id)</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-0ffec57995014fdd562d77d104186508b6463be91485f2c779bb4d3440b2474d">+118/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_qodo_parser.py</strong><dd><code>Comprehensive tests for Qodo parser</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_qodo_parser.py

<ul><li>46 comprehensive tests for qodo_parser module<br> <li> Tests for <code>parse_qodo_comment()</code> dispatcher logic<br> <li> Tests for <code>parse_improve_comment()</code> with real HTML fixtures<br> <li> Tests for <code>parse_review_comment()</code> with real HTML fixtures<br> <li> Tests for <code>_extract_path_from_link_text()</code> path extraction helper</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-db192828be788235576c690af7f1b764cb519689fceeecb115e33d51bd66487f">+541/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_get_all_reviews.py</strong><dd><code>Tests for Qodo issue comment fetching</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_get_all_reviews.py

<ul><li>Add test for <code>get_thread_key()</code> with issue comment suggestion composite <br>key<br> <li> Add 6 tests for <code>fetch_qodo_issue_comments()</code> covering filtering, <br>mapping, and error handling<br> <li> Tests verify author filtering, suggestion-to-thread mapping, API <br>failures</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-bcdbb49c2553d71205c3c0fdae98427dc0ec3d404bb60ebb096b2022c65ae851">+107/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_post_review_replies.py</strong><dd><code>Tests for issue comment posting and batch handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_post_review_replies.py

<ul><li>Add 3 tests for <code>post_issue_comment()</code> covering success, failure, and <br>truncation<br> <li> Add 4 tests for <code>build_issue_comment_reply()</code> covering formatting and <br>escaping<br> <li> Add 3 tests for issue comment handling in <code>run()</code> covering batch posting <br>and pending skipping<br> <li> Tests verify batch posting per comment ID and no thread resolution</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-3202dc39ff1a245242a16fadf1fde9407d2ccccac7901baa90f5dd342354e873">+239/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>review-handler.md</strong><dd><code>Document Qodo issue comment suggestion workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

plugins/myk-github/commands/review-handler.md

<ul><li>Document Phase 1 auto-fetching of Qodo issue comments<br> <li> Document Phase 6 batch posting of summary replies for issue comment <br>suggestions<br> <li> Clarify that issue comments are not posted as review threads<br> <li> Explain that no thread resolution occurs for issue comment suggestions</ul>

</details>

  </td>
  <td><a href="https://github.com/myk-org/claude-code-config/pull/115/files#diff-220f6bf1e4e5a0008e5b98ad3ae434442afbd2e4b5b7f19e4a64ada912f9590c">+13/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Automatic fetching, parsing, deduplication and batching of Qodo issue-comment suggestions; consolidated replies posted per original comment.

* **Documentation**
  * Workflow docs updated to describe the new batch-posting behavior for Qodo issue-comment suggestions.

* **Tests**
  * Extensive new tests covering Qodo comment parsing, fetching, deduplication, batching, posting, and related edge cases.

* **Chores**
  * Added two entries to the project ignore list.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -67,6 +67,7 @@
 !tests/test_post_review_replies.py
 !tests/test_review_db.py
 !tests/test_rule_enforcer.py
+!tests/test_qodo_parser.py
```
[the remark this answers is not in the document]
Comment (myakove) [reply .gitignore]: Fixed alphabetical order: moved `!tests/test_qodo_parser.py` between `test_post_review_replies.py` and `test_review_db.py`.

File: myk_claude_tools/reviews/post.py [hunk 2]
```diff
@@ -331,6 +401,21 @@ def run(json_path: str) -> None:
...
+                    already_posted_count += 1
+                    eprint(f"Skipping {category}[{i}] ({path}): issue comment suggestion already posted")
+                    continue
+                ic_id = thread_data.get("issue_comment_id")
+                if ic_id is not None:
+                    # Store category and index for JSON updates
+                    thread_data["_cat"] = category
+                    thread_data["_idx"] = i
+                    issue_comment_groups.setdefault(ic_id, []).append(thread_data)
+                continue
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_claude_tools/reviews/post.py]: Refactored to use `(category, index, thread_data)` tuples instead of mutating thread_data with `_cat`/`_idx`.

File: myk_claude_tools/reviews/qodo_parser.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+_REVIEW_FOCUS_RE = re.compile(
+    r"Recommended focus areas for review",
+    re.IGNORECASE,
+)
+
+_REVIEW_DETAILS_RE = re.compile(
+    r"<details>\s*<summary>\s*(?P<summary>.*?)</summary>\s*(?P<content>.*?)</details>",
+    re.DOTALL,
+)
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_claude_tools/reviews/qodo_parser.py]: Skipped: Works correctly today; complexity not warranted for hypothetical change

File: myk_claude_tools/reviews/qodo_parser.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            continue
+
+        # --- Description ---
+        # Text after </a> and before </summary> (within summary block)
+        desc: str | None = None
+        desc_match = re.search(r"</a>\s*\n?\s*(?P<desc>.+)", summary, re.DOTALL)
+        if desc_match:
+            raw_desc = desc_match.group("desc").strip()
+            # Clean HTML tags
+            raw_desc = re.sub(r"<[^>]+>", "", raw_desc).strip()
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_claude_tools/reviews/qodo_parser.py]: Skipped: Minimal risk; would require new dependency

**Comment by myakove** [comment]
### Review of [Qodo suggestions](https://github.com/myk-org/claude-code-config/pull/115#issuecomment-3901693855)

| # | Path | Status | Reply |
|---|------|--------|-------|
| 1 | `myk_claude_tools/reviews/post.py` | addressed | Changed to use `--input -` with JSON payload via stdin, which is more robust than `--raw-field` f... |
| 2 | `myk_claude_tools/reviews/post.py` | addressed | Added `int()` conversion with `try/except (TypeError, ValueError)` guard for `issue_comment_id` b... |
| 3 | `myk_claude_tools/reviews/fetch.py` | addressed | Added `comment_id is None` check and `int()` conversion with error handling in `fetch_qodo_issue_... |
| 4 | `myk_claude_tools/reviews/fetch.py` | addressed | Added `isinstance(comments, list)` validation after the None check in `fetch_qodo_issue_comments()`. |

**Comment by myakove** [comment]
### Review of [Qodo suggestions](https://github.com/myk-org/claude-code-config/pull/115#issuecomment-3901720451)

| # | Path | Status | Reply |
|---|------|--------|-------|
| 1 | `—` | addressed | Changed `post_issue_comment()` to use `--input -` with JSON payload via stdin. Same fix as item 0. |
| 2 | `—` | addressed | Added `comment_id` None check and `int()` validation. Same fix as item 5. |
| 3 | `—` | not_addressed | General awareness concern noted. The parser is tested against real Qodo data from multiple PRs an... |

File: myk_claude_tools/reviews/qodo_parser.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+            "path": None,
+            "line": None,
+            "end_line": None,
+            "diff": None,
+            "importance": None,
+            "impact": None,
+            "body": "\n".join(body_parts) if body_parts else title,
+            "qodo_type": "review",
+        })
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_claude_tools/reviews/qodo_parser.py]: Added comment documenting that /review focus areas use SHA-based diff anchors, so path extraction is intentionally omitted.

**Comment by myakove** [comment]
### Review of [Qodo suggestions](https://github.com/myk-org/claude-code-config/pull/115#issuecomment-3901691907)

| # | Path | Status | Reply |
|---|------|--------|-------|
| 1 | `—` | addressed | Already addressed: changed to `--input -` with JSON payload via stdin. |
| 2 | `—` | not_addressed | Already acknowledged: parser is tested against real Qodo data with defensive measures. |
| 3 | `—` | addressed | Already addressed: refactored to tuple-based tracking. |
